web/src/components/settings.tsx

```diff
@@ -0,0 +1,326 @@
+import { ChangeEvent } from "react";
+import TextareaAutosize from "react-textarea-autosize";
+
+import type { Parseable, LLMSettings, Entry, Mode } from "../types";
+import { SectionHeader, NumberInput, Slider } from "../components/html";
+
+type LLMSettingsParsers = {
+  [key: string]:
+    | ((v: number | undefined) => any)
+    | ((v: string | undefined) => any)
+    | ((v: object | undefined) => any);
+};
+
+const DEFAULT_PROMPTS = {
+  context:
+    "You are a helpful assistant knowledgeable about AI Alignment and Safety. " +
+    'Please give a clear and coherent answer to the user\'s questions.(written after "Q:") ' +
+    "using the following sources. Each source is labeled with a letter. Feel free to " +
+    "use the sources in any order, and try to use multiple sources in your answers.\n\n",
+  history:
+    "\n\n" +
+    'Before the question ("Q: "), there will be a history of previous questions and answers. ' +
+    "These sources only apply to the last question. any sources used in previous answers " +
+    "are invalid.",
+  question:
+    "In your answer, please cite any claims you make back to each source " +
+    "using the format: [a], [b], etc. If you use multiple sources to make a claim " +
+    'cite all of them. For example: "AGI is concerning [c, d, e]."\n\n',
+  modes: {
+    default: "",
+    concise:
+      "Answer very concisely, getting to the crux of the matter in as " +
+      "few words as possible. Limit your answer to 1-2 sentences.\n\n",
+    rookie:
+      "This user is new to the field of AI Alignment and Safety - don't " +
+      "assume they know any technical terms or jargon. Still give a complete answer " +
+      "without patronizing the user, but take any extra time needed to " +
+      "explain new concepts or to illustrate your answer with examples. " +
+      "Put extra effort into explaining the intuition behind concepts " +
+      "rather than just giving a formal definition.\n\n",
+  },
+};
+export const MODELS = {
+  "gpt-3.5-turbo": { maxNumTokens: 4095, topKBlocks: 10 },
+  "gpt-3.5-turbo-16k": { maxNumTokens: 16385, topKBlocks: 30 },
+  "gpt-4": { maxNumTokens: 8192, topKBlocks: 20 },
+  /* 'gpt-4-32k': {maxNumTokens: 32768, topKBlocks: 30}, */
+};
+export const ENCODERS = ["cl100k_base"];
+
+/** Update the given `obj` so that it has `val` at the given path.
```

Comment:
a bit of magic starts here - the idea is for each settings field to have a defined parser that will also return a default value if provided with an undefined value. Most of the magic is to handle the prompts, which are nested.
I considered just having separate keys for the prompt parts, but that would require a lot of messing around in other places, so I just kept the complexity here